testdata/p4_16_samples/v1model-special-ops-bmv2.p4

```diff
@@ -0,0 +1,470 @@
+{
+    const bit<16> ETHERTYPE_IPV4 = 16w0x0800;
+
+    state start {
+        transition parse_ethernet;
+    }
+    state parse_ethernet {
+        packet.extract(hdr.ethernet);
+        transition select(hdr.ethernet.etherType) {
+            ETHERTYPE_IPV4: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        packet.extract(hdr.ipv4);
+        transition accept;
+    }
+}
+
+#ifdef ENABLE_DEBUG_TABLES
+control debug_std_meta(in standard_metadata_t standard_metadata)
+{
+    table dbg_table {
+        key = {
+            // This is a complete list of fields inside of the struct
+            // standard_metadata_t as of the 2018-Sep-01 version of
+            // p4c in the file p4c/p4include/v1model.p4.
+
+            // parser_error is commented out because the p4c back end
+            // for bmv2 as of that date gives an error if you include
+            // a field of type 'error' in a table key.
+            standard_metadata.ingress_port : exact;
+            standard_metadata.egress_spec : exact;
+            standard_metadata.egress_port : exact;
+            standard_metadata.clone_spec : exact;
+            standard_metadata.instance_type : exact;
+            standard_metadata.drop : exact;
+            standard_metadata.recirculate_port : exact;
+            standard_metadata.packet_length : exact;
+            standard_metadata.enq_timestamp : exact;
+            standard_metadata.enq_qdepth : exact;
+            standard_metadata.deq_timedelta : exact;
+            standard_metadata.deq_qdepth : exact;
+            standard_metadata.ingress_global_timestamp : exact;
+            standard_metadata.egress_global_timestamp : exact;
+            standard_metadata.lf_field_list : exact;
+            standard_metadata.mcast_grp : exact;
+            standard_metadata.resubmit_flag : exact;
+            standard_metadata.egress_rid : exact;
+            standard_metadata.checksum_error : exact;
+            standard_metadata.recirculate_flag : exact;
+            //standard_metadata.parser_error : exact;
+        }
+        actions = { NoAction; }
+        const default_action = NoAction();
+    }
+    apply {
+        dbg_table.apply();
+    }
+}
+
+control my_debug_1(in headers_t hdr, in meta_t meta)
+{
+    table dbg_table {
+        key = {
+            hdr.ipv4.dstAddr : exact;
+            meta.fwd.l2ptr : exact;
+            meta.fwd.out_bd : exact;
+        }
+        actions = { NoAction; }
+        const default_action = NoAction();
+    }
+    apply {
+        dbg_table.apply();
+    }
+}
+#endif  // ENABLE_DEBUG_TABLES
+
+
+control fill_ipv4_address(out bit<32> ipv4_address,
+                          in bit<8> byte0,    // most significant byte
+                          in bit<8> byte1,
+                          in bit<8> byte2,
+                          in bit<8> byte3)    // least significant byte
+{
+    apply {
+        ipv4_address = (((bit<32>) byte0 << 24) |
```

Comment:
must easier to use concatenation `++`

Reply:
I am making that change and updating this PR soon.  Doh!  Can't believe I forgot about concatenation while writing out that expression.